### DISCUSSION TITLE
How can I update my embedded Linux program with a removable disk?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#14776] [Q&A] [answered]
I made an embedded Linux program using Avalonia.

I plan to add some useful features to the program to make it even better.

When a new version of a program is released (for example, because of NuGet package upgrades, bug fixes, new features, etc.), if I want to upgrade a new program for an embedded device, I always have to take the device back to my office, plug it into my computer, it is very troublesome to transfer the released new program through FTP and wait for a series of complicated steps.

I noticed that most embedded devices on the market support upgrades through removeable disk. Insert a removeable disk containing the upgrade package into the USB port of the embedded device, and then click a button on the screen to automatically upgrade. This feels very useful.

But I don't know how to implement this function for Avalonia, or even whether it can be implemented.

An immature idea I have is to create an upgrade program in addition to the main program. When the removeable disk is inserted into the USB port of the embedded device and click upgrade, change the Linux boot auto-start program to the upgrade program and automatically restart the system , and then the upgrade program completes the copy and paste of files for upgrade. When the upgrade is completed and confirmed (for example, md5 verification of the file), then re-modify the Linux boot auto-start program back to the main program and restart the system to complete the upgrade.

I'm not sure if implementing this has more to do with Avalonia or Linux, so would like to hear everyone's experiences and thoughts.

**Comment by thevortexcloud** [comment, marked as answer]
> An immature idea I have is to create an upgrade program in addition to the main program.

This is pretty much how most old/simple programs/devices handle updates. Some programs do allow for overwriting files on disk while a program is active, but most don't. .NET certainly does not since it locks all the files. 

There are a few different ways to handle this. The simplest would be to:

1. Copy/download the update files to a staging area on the device.
2. Have your main app start your update app.
3. Have your updater app check if there is any updates available, and if there is wait for your main app to exit.
4. Have the updater app replace your main app (ideally with some sort of transactional IO so if the update fails it can be rolled back).
5. Cleanup your update files
6. Restart your main app somehow

Fancier update systems do A/B updates. Which means there will always be at least two copies of the app on the device at all times. If one of the copies stops working for what ever reason, the system will switch to the other copy. This also means you have some protection against failed updates as you would leapfrog style update each copy. EG Copy A is updated,  and copy B is left intact. Shortly after a new update is released and copy B is updated and copy A is left alone. Each time the app is updated you switch which copy you boot from.

The other big advantage of A/B updates is you can skip having a dedicated upgrade program and you can do updates "live" without exiting your main app. Which is pretty much how androids update system works. You only have to reboot at the end to switch to the updated copy of the OS.

**Reply by kyuranger** [reply]
Thank you very much for your suggestion. A/B updates are indeed a good way.

**Reply by thevortexcloud** [reply]
You are welcome. If you do decide to go down the A/B route, the general practice in embedded scenarios is to use an immutable file system to prevent accidental/intentional file changes to the active and/or inactive version of the program. It also make updates easier, since it means you just destroy the entire existing install without worrying about what is already on the device. The file system will only ever be writeable during updates. However things can get fairly complicated at that point since you may not have anywhere to write files to during normal operation. But that may not be a problem for you if you don't write files to any location. There are works arounds for that on Linux using things like [overlays](https://wiki.archlinux.org/title/Overlay_filesystem) or just making a separate data partition that is separate from your app files. 

Anyway, you don't have to do any of that if you don't want to.
